modules/stochastic_tools/src/surrogates/PolynomialChaosTrainer.C

```diff
@@ -44,53 +49,139 @@ PolynomialChaosTrainer::PolynomialChaosTrainer(const InputParameters & parameter
 {
-  _coeff.clear();
-  _coeff.resize(_ncoeff, 0.0);
+  _coeff.assign(_ncoeff, 0.0);
+
+  if (_rtype == 1)
+  {
+    _matrix.resize(_ncoeff, _ncoeff);
+    _rhs.resize(_ncoeff);
+    for (auto & calc : _calculators)
+      calc->initializeCalculator();
+    _r_sum = 0.0;
+  }
 }
 
 void
 PolynomialChaosTrainer::train()
 {
-  DenseMatrix<Real> poly_val(_ndim, _order);
-
   // Evaluate polynomials to avoid duplication
+  DenseMatrix<Real> poly_val(_ndim, _order);
   for (unsigned int d = 0; d < _ndim; ++d)
     for (unsigned int i = 0; i < _order; ++i)
-      poly_val(d, i) = _poly[d]->compute(i, _predictor_row[d]);
+      poly_val(d, i) = _poly[d]->compute(i, _predictor_row[d], /*normalize=*/_rtype == 0);
+
+  // Evaluate multi-dimensional polynomials
+  std::vector<Real> basis(_ncoeff, 1.0);
+  for (const auto i : make_range(_ncoeff))
+    for (const auto d : make_range(_ndim))
+      basis[i] *= poly_val(d, _tuple[i][d]);
 
-  // Loop over coefficients
-  for (std::size_t i = 0; i < _ncoeff; ++i)
+  // For integration
+  if (_rtype == 0)
   {
-    Real val = *_rval;
-    // Loop over parameters
-    for (std::size_t d = 0; d < _ndim; ++d)
-      val *= poly_val(d, _tuple[i][d]);
-
-    if (_quad_sampler)
-      val *= _quad_sampler->getQuadratureWeight(_row);
-    _coeff[i] += val;
+    const Real fact = (*_rval) * (_quad_sampler ? _quad_sampler->getQuadratureWeight(_row) : 1.0);
+    for (const auto i : make_range(_ncoeff))
+      _coeff[i] += fact * basis[i];
+  }
+  // For least-squares
+  else
+  {
+    // Loop over coefficients
+    for (const auto i : make_range(_ncoeff))
+    {
+      // Matrix is symmetric, so we'll add the upper diagonal later
+      for (const auto j : make_range(i + 1))
+        _matrix(i, j) += basis[i] * basis[j];
+      _rhs(i) += basis[i] * (*_rval);
+
+      for (unsigned int c = i * 3; c < (i + 1) * 3; ++c)
+        _calculators[c]->updateCalculator(basis[i]);
+    }
+    _r_sum += (*_rval);
   }
 }
 
 void
 PolynomialChaosTrainer::postTrain()
 {
-  gatherSum(_coeff);
+  if (_rtype == 0)
+  {
+    gatherSum(_coeff);
+    if (!_quad_sampler)
+      for (std::size_t i = 0; i < _ncoeff; ++i)
+        _coeff[i] /= getCurrentSampleSize();
+  }
+  else
+  {
+    gatherSum(_matrix.get_values());
+    gatherSum(_rhs.get_values());
+    for (auto & calc : _calculators)
+      calc->finalizeCalculator(true);
+    gatherSum(_r_sum);
+
+    std::vector<Real> mu(_ncoeff);
+    std::vector<Real> sig(_ncoeff);
+    std::vector<Real> sum_pf(_ncoeff);
+    for (const auto i : make_range(_ncoeff))
+    {
+      mu[i] = i > 0 ? _calculators[3 * i]->getValue() : 0.0;
+      sig[i] = i > 0 ? _calculators[3 * i + 1]->getValue() : 1.0;
+      sum_pf[i] = _calculators[3 * i + 2]->getValue();
+    }
+
+    const Real n = getCurrentSampleSize();
+    for (const auto i : make_range(_ncoeff))
+    {
+      for (const auto j : make_range(i + 1))
+      {
+        _matrix(i, j) -= (mu[j] * sum_pf[i] + mu[i] * sum_pf[j]);
```

Comment:
Do you think standardization needs to be mandatory? I would make sense, just asking if you see an application in the future where someone wants to avoid standardization.